From 8636a230c312a47692909775abad826e3b953bf1 Mon Sep 17 00:00:00 2001
From: chdugw_auto <jkulkarni@maxlinear.com>
Date: Tue, 10 Aug 2021 13:36:59 +0300
Subject: [PATCH] Merge pull request #3458 in SW_UGW/linux from
 feature/DRVLIB_SW-4166-polling-func-to-check-xgmac-rx_fifo_error-and-reset-adap-layer
 to master

* commit 'a20a281f62abf6351aaddd53d6e0c0de4e2b15f6':
  DRVLIB_SW-4166: Sanity Check if RxFifo Error happens reset switch adaption layer
---
 .../ethernet/lantiq/switch-api/gswip_dev/gsw_dev.c | 61 ++++++++++++++++++++++
 .../ethernet/lantiq/switch-api/mac/xgmac_common.h  |  9 ++++
 2 files changed, 70 insertions(+)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gswip_dev/gsw_dev.c b/drivers/net/ethernet/lantiq/switch-api/gswip_dev/gsw_dev.c
index 6c660316e2bf..bbd73ee25a9c 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gswip_dev/gsw_dev.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gswip_dev/gsw_dev.c
@@ -15,6 +15,8 @@
 #include <linux/of.h>
 #include <linux/reset.h>
 #include <xgmac_common.h>
+#include <xgmac.h>
+#include <gswss_mac_api.h>
 #include <gswss_api.h>
 #include <gsw_flow_core.h>
 
@@ -279,6 +281,10 @@ static int gsw_add_macdev(struct gsw_cell *gsw_dev_cell, u32 devid)
 	/* Retrieve the speedset */
 	ret = of_property_read_u32(gsw_dev_cell->of_node, "speed",
 				   &mac_pdata->phy_speed);
+
+	/* Logic to Reset Adaption layer if RxFifo Error happens in MAC */
+	of_property_read_u32(gsw_dev_cell->of_node, "rxfifo_err_thresh",
+			     &mac_pdata->rxfifo_err.rxfifo_err_threshold);
 #else
 
 	if (gsw_dev[devid].prod_id & PRX300) {
@@ -727,6 +733,58 @@ static int gsw_add_devices(struct device *parent, int id,
 	return ret;
 }
 
+/* This API will get executed by workqueue
+ * if per MAC rxfifo_err exceeds rxfifo_err_threshold,
+ * it resets corresponding MAC adaption layer
+ */
+static int rxfifo_err_poll(struct delayed_work *work)
+{
+	struct mac_rxfifo_poll *rxfifo_err =
+		container_of(work, struct mac_rxfifo_poll, work);
+	u32 devid = 0;
+	int mac_cnt = gsw_get_mac_subifcnt(devid);
+	int i = 0;
+	struct mac_ops *ops;
+	u32 val_lo = 0, val_hi = 0;
+	u64 val = 0;
+	int err_cnt = 0;
+
+	for (i = 0; i < mac_cnt; i++) {
+		ops = gsw_get_mac_ops(devid, MAC_2 + i);
+
+		ops->xgmac_reg_rd(ops, MMC_RXFIFOOVERFLOW_LO, &val_lo);
+		ops->xgmac_reg_rd(ops, MMC_RXFIFOOVERFLOW_LO + 4, &val_hi);
+		val = ((u64)val_hi << 32) | val_lo;
+
+		if (!rxfifo_err->rxfifo_err_cnt[i]) {
+			rxfifo_err->rxfifo_err_cnt[i] = val;
+			err_cnt = val;
+		} else {
+			err_cnt = val - rxfifo_err->rxfifo_err_cnt[i];
+		}
+
+		if (err_cnt > rxfifo_err->rxfifo_err_threshold) {
+			gswss_adap_reset(ops, 1);
+			rxfifo_err->rxfifo_err_cnt[i] = val;
+		}
+	}
+
+	/* reschedule to check later after 1 sec*/
+	schedule_delayed_work(&rxfifo_err->work, msecs_to_jiffies(1000));
+
+	return 1;
+}
+
+static void cfg_rxfifo_err_poll(struct platform_device *pdev, int devid)
+{
+	struct mac_ops *ops = gsw_get_mac_ops(devid, MAC_2);
+	struct mac_prv_data *pdata = GET_MAC_PDATA(ops);
+
+	/* Init and schedule RxFifo Error Poll for all MAC's in every 1 sec */
+	INIT_DELAYED_WORK(&pdata->rxfifo_err.work, (work_func_t)rxfifo_err_poll);
+	schedule_delayed_work(&pdata->rxfifo_err.work, msecs_to_jiffies(1000));
+}
+
 static int gsw_dev_probe(struct platform_device *pdev)
 {
 	u32 devid = 0;
@@ -788,6 +846,9 @@ static int gsw_dev_probe(struct platform_device *pdev)
 	if (ret == 0 && prod_id == PRX300)
 		cfg_ptp_mux(pdev, devid);
 
+	if (ret == 0 && prod_id != GRX500)
+		cfg_rxfifo_err_poll(pdev, devid);
+
 	return ret;
 }
 
diff --git a/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_common.h b/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_common.h
index 6e83b1f5780b..cbd591fb5761 100644
--- a/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_common.h
+++ b/drivers/net/ethernet/lantiq/switch-api/mac/xgmac_common.h
@@ -340,6 +340,12 @@ struct mac_irq_hdl {
 	void *param;
 };
 
+/* Structure to poll for Rx Fifo Error and set adaptation layer */
+struct mac_rxfifo_poll {
+	struct delayed_work work;
+	u64 rxfifo_err_cnt[MAC_LAST];
+	int rxfifo_err_threshold;
+};
 
 /* do forward declaration of private data structure */
 struct mac_prv_data;
@@ -567,6 +573,9 @@ struct mac_prv_data {
 
 	/* MAC IRQ handler */
 	struct mac_irq_hdl *irq_hdl;
+
+	/* MAC RxFifo Err Poll */
+	struct mac_rxfifo_poll rxfifo_err;
 };
 
 extern struct mac_prv_data prv_data[10];
